src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -277,10 +276,29 @@ public async Task ResolveBundleReferences(EntryComponent entry, Dictionary<strin
                     continue;
                 }
 
+                string versionId = string.Empty;
+
+                if (reference.Reference.Contains("/_history/", StringComparison.OrdinalIgnoreCase))
+                {
+                    string[] versionedReference = reference.Reference.Split("/_history/");
+                    reference.Reference = versionedReference[0];
+                    versionId = versionedReference[1];
+                }
+
                 // Checks to see if this reference has already been assigned an Id
                 if (referenceIdDictionary.TryGetValue(reference.Reference, out var referenceInformation))
                 {
-                    reference.Reference = $"{referenceInformation.resourceType}/{referenceInformation.resourceId}";
+                    if (string.IsNullOrWhiteSpace(versionId))
```

Comment:
Probably, we can get rid of extra if statement
if (referenceInformation.versionId == versionId)
 {
   reference.Reference = $"{referenceInformation.resourceType}/{referenceInformation.resourceId}/_history/{referenceInformation.versionId}";
 }
else
 {
   reference.Reference = $"{referenceInformation.resourceType}/{referenceInformation.resourceId}";
 }

Reply:
We need to explicitly check for nullorempty and there for it is there. Because all reference may not have versionId. I will clean extra if though.